prefork/prefork.go

```diff
@@ -0,0 +1,179 @@
+package prefork
+
+import (
+	"flag"
+	"net"
+	"os"
+	"os/exec"
+	"runtime"
+
+	"github.com/valyala/fasthttp"
+	"github.com/valyala/fasthttp/reuseport"
+)
+
+const preforkChildFlag = "prefork-child"
+const defaultNetwork = "tcp4"
+
+var child bool
+
+// Prefork implements fasthttp server prefork
+//
+// Preforks master process (with all cores) between several child processes
+// increases performance significantly, because Go doesn't have to share
+// and manage memory between cores
+//
+// WARNING: using prefork prevents the use of any global state!
+// Things like in-memory caches won't work.
+type Prefork struct {
+	Addr string
+
+	// The network must be "tcp", "tcp4" or "tcp6".
+	//
+	// By default is "tcp4"
+	Network string
+
+	// Flag to use a listener with reuseport, if not a File Listener will be used
+	//
+	// It's disabled by default
+	Reuseport bool
+
+	ServeFunc         func(ln net.Listener) error
+	ServeTLSFunc      func(ln net.Listener, certFile, keyFile string) error
+	ServeTLSEmbedFunc func(ln net.Listener, certData, keyData []byte) error
+
+	ln    net.Listener
+	files []*os.File
+}
+
+func init() { // nolint:gochecknoinits
+	flag.BoolVar(&child, preforkChildFlag, false, "is child proc")
+}
+
+// New wraps the fasthttp server to run with prefork processes
+func New(s *fasthttp.Server) *Prefork {
+	return &Prefork{
+		Network:           defaultNetwork,
+		ServeFunc:         s.Serve,
+		ServeTLSFunc:      s.ServeTLS,
+		ServeTLSEmbedFunc: s.ServeTLSEmbed,
+	}
+}
+
+func (p *Prefork) listen(addr string) (net.Listener, error) {
+	p.Addr = addr
+
+	runtime.GOMAXPROCS(1)
+
+	if p.Network == "" {
+		p.Network = defaultNetwork
+	}
+
+	if p.Reuseport {
+		return reuseport.Listen(p.Network, p.Addr)
+	}
+
+	return net.FileListener(os.NewFile(3, ""))
+}
+
+func (p *Prefork) setTCPListenerFiles(addr string) error {
+	p.Addr = addr
+
+	tcpAddr, err := net.ResolveTCPAddr(p.Network, p.Addr)
+	if err != nil {
+		return err
+	}
+
+	tcplistener, err := net.ListenTCP(p.Network, tcpAddr)
+	if err != nil {
+		return err
+	}
+
+	p.ln = tcplistener
+
+	fl, err := tcplistener.File()
+	if err != nil {
+		return err
+	}
+
+	p.files = []*os.File{fl}
+
+	return nil
+}
+
+func (p *Prefork) prefork(addr string) error {
+	strCmd := os.Args[0]
+	chErr := make(chan error, 1)
+
+	if !p.Reuseport {
+		if err := p.setTCPListenerFiles(addr); err != nil {
+			return err
+		}
+
+		defer p.ln.Close()
+	}
+
+	for i := 0; i < runtime.GOMAXPROCS(0); i++ {
+		cmd := exec.Command(strCmd, append(os.Args[1:], "-"+preforkChildFlag)...)
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		cmd.ExtraFiles = p.files
+
+		go func() {
+			chErr <- cmd.Run()
+		}()
+	}
+
+	return <-chErr
+}
+
+// ListenAndServe serves HTTP requests from the given TCP addr
+func (p *Prefork) ListenAndServe(addr string) error {
+	if child {
```

Comment:
Replace this with:
```go
if isChild() {
```
And
```
func isChild() bool {
	for _, arg := range os.Args {
		if arg == preforkChildFlag {
			return true
		}
	}
	return false
}
```
And just make `preforkChildFlag = "-prefork-child"` etc...

Reply:
Yes, i didn't understand it well before.
Fixed!